core/src/main/java/org/apache/calcite/rel/rules/JoinConditionOrExpansionRule.java

```diff
@@ -0,0 +1,297 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelOptUtil;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.tools.RelBuilder;
+
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Planner rule that matches a
+ * {@link org.apache.calcite.rel.core.Join}
+ * and expands OR clauses in join conditions.
+ *
+ * <p>this rule would expand the OR condition into
+ * two separate join conditions, allowing the optimizer
+ * to handle these conditions more effectively.
+ */
+@Value.Enclosing
+public class JoinConditionOrExpansionRule
+    extends RelRule<JoinConditionOrExpansionRule.Config>
+    implements TransformationRule {
+
+  /** Creates an JoinConditionExpansionOrRule. */
+  protected JoinConditionOrExpansionRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    Join join = call.rel(0);
+    RelBuilder relBuilder = call.builder();
+    List<RexNode> orConds = RelOptUtil.disjunctions(join.getCondition());
+
+    if (orConds.size() <= 1) {
+      return;
+    }
+
+    RelNode converted;
+    switch (join.getJoinType()) {
+    case INNER:
+      converted = expandInnerJoin(join, orConds, relBuilder);
+      break;
+    case SEMI:
+      converted =
+          expandSemiOrAntiJoin(join, JoinRelType.SEMI, orConds, relBuilder);
+      break;
+    case ANTI:
+      converted =
+          expandSemiOrAntiJoin(join, JoinRelType.ANTI, orConds, relBuilder);
+      break;
+    case LEFT:
+      converted = expandLeftJoin(join, orConds, relBuilder);
+      break;
+    case RIGHT:
+      converted = expandRightJoin(join, orConds, relBuilder);
+      break;
+    case FULL:
+      converted = expandFullJoin(join, orConds, relBuilder);
+      break;
+    default:
+      return;
+    }
+    call.transformTo(converted);
+  }
+
+  /**
+   * This method will make the following conversions.
+   *
+   * <p>Project[*]
+   *    └── Join[OR(id=id0, age=age0), left]
+   *        ├── TableScan[tbl]
+   *        └── TableScan[tbl]
+   *
+   * <p>into
+   *
+   * <p>Project[*]
+   *    └── UnionAll
+   *        ├── HashJoin[id=id0, inner]
```

Comment:
technically I don't think these are hash joins, they are all logical joins, aren't they?

Reply:
Yes, you're right!

Reply:
Done